src/theory/uf/proof_equality_engine.cpp

```diff
@@ -0,0 +1,762 @@
+    {
+      // we do not process this fact if it already holds
+      return false;
+    }
+    // buffer the steps in the fact proof generator
+    const std::vector<std::pair<Node, ProofStep>>& steps = psb.getSteps();
+    for (const std::pair<Node, ProofStep>& step : steps)
+    {
+      d_factPg.addStep(step.first, step.second);
+    }
+    // add lazy step to proof
+    d_proof.addLazyStep(lit, &d_factPg, false);
+  }
+  // second, assert it to the equality engine
+  return assertFactInternal(atom, polarity, exp);
+}
+
+bool ProofEqEngine::assertFact(Node lit, Node exp, ProofGenerator* pg)
+{
+  Trace("pfee") << "pfee::assertFact " << lit << ", exp = " << exp
+                << " via generator" << std::endl;
+  Node atom = lit.getKind() == NOT ? lit[0] : lit;
+  bool polarity = lit.getKind() != NOT;
+  if (d_pfEnabled)
+  {
+    if (holds(atom, polarity))
+    {
+      // we do not process this fact if it already holds
+      return false;
+    }
+    // note the proof generator is responsible for remembering the explanation
+    d_proof.addLazyStep(lit, pg, false);
+  }
+  // second, assert it to the equality engine
+  return assertFactInternal(atom, polarity, exp);
+}
+
+TrustNode ProofEqEngine::assertConflict(Node lit)
+{
+  Trace("pfee") << "pfee::assertConflict " << lit << std::endl;
+  std::vector<TNode> assumps;
+  explainWithProof(lit, assumps, &d_proof);
+  if (d_pfEnabled)
+  {
+    // lit may not be equivalent to false, but should rewrite to false
+    if (lit != d_false)
+    {
+      Assert(Rewriter::rewrite(lit) == d_false)
+          << "pfee::assertConflict: conflict literal is not rewritable to "
+             "false";
+      std::vector<Node> exp;
+      exp.push_back(lit);
+      std::vector<Node> args;
+      if (!d_proof.addStep(d_false, PfRule::MACRO_SR_PRED_ELIM, exp, args))
+      {
+        Assert(false) << "pfee::assertConflict: failed conflict step";
+        return TrustNode::null();
+      }
+    }
+  }
+  return ensureProofForFact(
+      d_false, assumps, TrustNodeKind::CONFLICT, &d_proof);
+}
+
+TrustNode ProofEqEngine::assertConflict(PfRule id,
+                                        const std::vector<Node>& exp,
+                                        const std::vector<Node>& args)
+{
+  Trace("pfee") << "pfee::assertConflict " << id << ", exp = " << exp
+                << ", args = " << args << std::endl;
+  // conflict is same as proof of false
+  std::vector<Node> empVec;
+  return assertLemma(d_false, id, exp, empVec, args);
+}
+
+TrustNode ProofEqEngine::assertConflict(const std::vector<Node>& exp,
+                                        ProofStepBuffer& psb)
+{
+  Trace("pfee") << "pfee::assertConflict " << exp << " via buffer with "
+                << psb.getNumSteps() << " steps" << std::endl;
+  if (d_pfEnabled)
+  {
+    if (!d_proof.addSteps(psb))
+    {
+      return TrustNode::null();
+    }
+  }
+  std::vector<Node> empVec;
+  return assertLemmaInternal(d_false, exp, empVec, &d_proof);
```

Comment:
Say why we need to call the internal version rather than assertLemma. It's not clear looking at the two versions of assertConflict why this difference.